### PR TITLE
Issue #7 Add connection generation

### PR DESCRIPTION
**Pull request by mhirna**
Add connection generation for issue #7.

### CODE REVIEW

File: connection_generator.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+        _dataset_write_to_system_output()
+            Generates connections between dataset write and system outputs.
+
+        generate()
+            Generates all the needed connections for data dependency mapping graph.
+    """
+    def __init__(self, dataset_count, dataset_count_map, system_count, system_count_map, dataset_read_count,
+                 system_input_count, dataset_write_count, system_output_count, dataset_read_count_map,
+                 dataset_write_count_map, system_input_count_map, system_output_count_map):
```
Comment (kboyeldieu) [inline connection_generator.py]: That is a lot of parameters! Could these be split into multiple groups and moved into separate classes? eg what's dataset related in one DatasetStats class, what's System related in one SystemStats class.

File: connection_generator.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
+"""
+This module implements methods for generating random connections between nodes in a graph.
```
Comment (kboyeldieu) [inline connection_generator.py:2]: nit: remove this dot

File: connection_generator.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        """Generate group id for each element, based on number of element in group distribution.
+
+        Args:
+            element_count: Total number of elements.
+            element_count_map: Dictionary that maps element count in a group to number of groups with that count.
+
+        Returns:
+            Dictionary that maps group id to a list of element ids.
+        """
+        # Create element ids
```
Comment (kboyeldieu) [inline connection_generator.py]: nit: All comments should end with a dot ;)

File: connection_generator.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+class ConnectionGenerator:
+    """
+    A class to generate random connections between node ids, based on distribution maps.
+
+    ...
+
+    Attributes:
+        dataset_count: Integer of how many datasets are in a graph.
+        dataset_count_map: Dictionary that maps number of datasets in collection to count of dataset collections.
```
Comment (kboyeldieu) [inline connection_generator.py]: As there's no test, consider adding type for dictionaries (e.g. int:int or int:[int]) for the reader to better understand

File: connection_generator.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        elements_per_group = [i for i in element_count_map for _ in range(element_count_map[i])]
+
+        # Randomise element ids and group ids
+        random.shuffle(element_values)
+        random.shuffle(elements_per_group)
+
+        it = iter(element_values)
+
+        # Split element ids into chunks to get connections for each group
+        group_to_elements = {i + 1: set(islice(it, 0, elements_per_group[i])) for i in range(len(elements_per_group))}
```
Comment (kboyeldieu) [inline connection_generator.py]: Does this need to be an iter?

File: connection_generator.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        # Get number of elements in each group and remove groups with 0 elements
+        elements_per_group_1 = [i for i in element_1_count_map for j in range(element_1_count_map[i]) if i != 0]
+        elements_per_group_2 = [i for i in element_2_count_map for j in range(element_2_count_map[i]) if i != 0]
+        element_1_group_counter = {i + 1: elements_per_group_1[i] for i in range(len(elements_per_group_1))}
+        element_2_group_counter = {i + 1: elements_per_group_2[i] for i in range(len(elements_per_group_2))}
+
+        # Create connection dictionary
+        element_1_conn_element_2 = {i: set() for i in element_1_values}
+
+        # Loop until run out of elements in any group
```
Comment (kboyeldieu) [inline connection_generator.py]: nit: there seems to be a missing word: loop until [we,it?] runs out..

File: connection_generator.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        while element_1_values and element_2_values:
```
Comment (kboyeldieu) [inline connection_generator.py:153]: How long does it take for this while loop to finish?

File: connection_generator.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+                                                                            self.system_output_count,
+                                                                            self.dataset_write_count_map,
+                                                                            self.system_output_count_map)
+
+    def generate(self):
+        """Generate all connections for a graph."""
+        self._dataset_to_dataset_collection()
+        self._system_to_system_collection()
+        self._dataset_read_to_system_input()
+        self._dataset_write_to_system_output()
```
Comment (kboyeldieu) [inline connection_generator.py:204]: Clear and nice! 👍

**Review by kboyeldieu — CHANGES_REQUESTED** [review]

File: connection_generator.py (hunk 7 above)
Comment (mhirna) [reply connection_generator.py:153]: Less than a second for a given config.

File: connection_generator.py (hunk 5 above)
Comment (mhirna) [reply connection_generator.py]: It doesn't have to be an iter, could be a loop with list slicing and setting it as a value in a dictionary. This way just looks prettier to me.

File: config_params/connection_params.py [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class ConnectionParams:
+    def __init__(self, system_write_conn_count, system_read_conn_count):
```
Comment (kboyeldieu) [inline config_params/connection_params.py]: nit: is "conn" a popular diminutive for connection? If not consider simply using "connection" or even omitting it?

File: process_config.py [hunk 10]
```diff
@@ -1,25 +1,15 @@
...
-
-    Args:
-        config_map: string that contains the map
-        proba: boolean, true if values are probabilities, not counts.
-        enum: boolean, true if keys are enums, not integers.
 
-    Returns:
-        A dictionary of keys and values in converted format.
-    """
+def process_map(config_map, proba=False, enum=False):
```
Comment (kboyeldieu) [inline process_config.py:17]: Did you remove the docstring?

File: process_config.py [hunk 11]
```diff
@@ -1,25 +1,15 @@
-"""Graph Generation Config Reader
-
-This script reads the config from full_config.yaml file and processes it to be usable.
-It saves config to variables, and converts map strings to python dictionaries.
-"""
-
```
Comment (kboyeldieu) [inline process_config.py]: is this deletion intended?

File: connection_generator.py (hunk 5 above)
Comment (kboyeldieu) [reply connection_generator.py]: I mean "it", does it need to be an iter (rather than a list)?

File: config_params/connection_params.py (hunk 9 above)
Comment (mhirna) [reply config_params/connection_params.py]: I think it is popular. But as it was unclear to you, will fix it.

File: process_config.py (hunk 10 above)
Comment (mhirna) [reply process_config.py:17]: Yes, by accident. Same with module docstring.

**Review by kboyeldieu — APPROVED** [review]

**Pull request merged**